### PR TITLE
[Port] OD layers to Keras 3

### PR DESCRIPTION
**Pull request by ariG23498**
TODOs:

- [x] `roi_align.py`
- [x] `roi_generator.py`
- [x] `roi_pool.py`
- [x] `roi_sampler.py`
- [x] `rpn_label_encoder.py`

CC: @divyashreepathihalli

### CODE REVIEW

File: keras_cv/layers/object_detection/roi_align.py
Comment (ariG23498) [inline keras_cv/layers/object_detection/roi_align.py:1]: There are certain blockers that I need some guidance on:
- `with tf.name_scope("multilevel_crop_and_resize")`
- `tf.constant`
- `tf.math.divide_no_nan`

Comment (AakashKumarNain) [reply keras_cv/layers/object_detection/roi_align.py:1]: What's the blocker exactly?

Comment (ariG23498) [reply keras_cv/layers/object_detection/roi_align.py:1]: I meant -- how should I be porting the above mentioned APIs to Keras3.

Comment (AakashKumarNain) [reply keras_cv/layers/object_detection/roi_align.py:1]: I don't think you require name scoping. For `constant`, use KerasTensor? For third one, just write a custom divide function. Though we should add it in the core actually

Comment (ariG23498) [reply keras_cv/layers/object_detection/roi_align.py:1]: Thanks for the tips @AakashKumarNain

I am documenting what I did:
- Removed scopes
- Used `ops.convert_to_tensor` or used `ops.ones` multiplying a scalar to it
- Use `ops.cond` for the divide no nan

**Comment by ariG23498** [comment]
@divyashreepathihalli Could you check the port of `roi_generator.py`? The only thing to note is that the NMS produces `-1.0` instead of `0.0` in the bounding boxes (this has been taken care of in the tests).

**Comment by divyashreepathihalli** [comment]
Thanks @ariG23498! roi_align and roi_generator LGTM!!

**Review by divyashreepathihalli — APPROVED** [review]
LGTM! Thank you.

**Comment by divyashreepathihalli** [comment]
Thanks for the updates @ariG23498. There is one failing test
```
FAILED keras_cv/layers/object_detection/rpn_label_encoder_test.py::RpnLabelEncoderTest::test_rpn_label_encoder_multi_level - AssertionError: 
Not equal to tolerance rtol=1e-06, atol=1e-06
Mismatched value: a is different from b. 
not close where = (array([0]), array([0]))
not close lhs = [0.]
not close rhs = [1.]
not close dif = [1.]
not close tol = [2.e-06]
dtype = float64, shape = (2, 1)
Mismatched elements: 1 / 2 (50%)
Max absolute difference: 1.
Max relative difference: 1.
 x: array([[0.],
       [1.]])
 y: array([[1.],
       [1.]], dtype=float32)
       ```

**Review by AakashKumarNain — APPROVED** [review]

**Comment by ariG23498** [comment]
> Thanks for the updates @ariG23498. There is one failing test
> ```
> FAILED keras_cv/layers/object_detection/rpn_label_encoder_test.py::RpnLabelEncoderTest::test_rpn_label_encoder_multi_level - AssertionError: 
> Not equal to tolerance rtol=1e-06, atol=1e-06
> Mismatched value: a is different from b. 
> not close where = (array([0]), array([0]))
> not close lhs = [0.]
> not close rhs = [1.]
> not close dif = [1.]
> not close tol = [2.e-06]
> dtype = float64, shape = (2, 1)
> Mismatched elements: 1 / 2 (50%)
> Max absolute difference: 1.
> Max relative difference: 1.
>  x: array([[0.],
>        [1.]])
>  y: array([[1.],
>        [1.]], dtype=float32)
>        ```

Hey @divyashreepathihalli do you think this one has to do with the `keras.radom` API? I am unable to track the source of the failing test.

**Comment by ariG23498** [comment]
@divyashreepathihalli @AakashKumarNain I have created a gist to explain the failing test `test_rpn_label_encoder_multi_level`

[GitHub Gist](https://colab.research.google.com/gist/ariG23498/e8322c221b3c1ed666a7cf4af5f83f80/scratchpad.ipynb)

One can notice that the tests are flaky due to the radom API for `sampling.py`. I am unsure why the tests pass in the first place (maybe a PRNG issue). How do you think we should mitigate this?

**Comment by divyashreepathihalli** [comment]
@ariG23498 the randomAPI has changed and these tests can be fllaky. Please disable the test and add and issue and add a todo in the code.

**Comment by ariG23498** [comment]
> @ariG23498 the randomAPI has changed and these tests can be fllaky. Please disable the test and add and issue and add a todo in the code.

I have made the changes requested. I am also adding the link of the issue to this thread.

https://github.com/keras-team/keras-cv/issues/2336

File: keras_cv/layers/object_detection/roi_sampler_test.py [hunk 1]
```diff
@@ -62,57 +62,58 @@ def test_roi_sampler(self):
             np.min(sampled_gt_classes),
         )
 
-    def test_roi_sampler_small_threshold(self):
-        box_matcher = BoxMatcher(thresholds=[0.1], match_values=[-1, 1])
```
Comment (divyashreepathihalli) [inline keras_cv/layers/object_detection/roi_sampler_test.py:71]: Please add `self.skipTest("TODO: type message and issue number here")` right before  `box_matcher = BoxMatcher(thresholds=[0.1], match_values=[-1, 1])`

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

File: keras_cv/layers/object_detection/roi_sampler_test.py (hunk 1 above)
Comment (ariG23498) [reply keras_cv/layers/object_detection/roi_sampler_test.py:71]: Got it! Thanks for the help.

**Comment by divyashreepathihalli** [comment]
Please rebase for the fix!

**Review by divyashreepathihalli — APPROVED** [review]

**Comment by divyashreepathihalli** [comment]
Great work on this @ariG23498! this work is very impactful! Thank you!!

**Pull request merged**